### DISCUSSION TITLE
Decoupling data querying from panel plugin

### DISCUSSION THREAD

**Discussion by eunicorn** [perses/perses#956] [Design Discussions]
**BACKGROUND**
Currently there are three parts to building a panel plugin. You need to define the panel component, query editor component and options editor component. The panel component is responsible for running the query and displaying the data. For example, TimeSeriesChartPanel calls `useTimeSeriesQueries` and renders the chart. This ties data querying and panel plugin together, which isn't ideal for several reasons.

1. A lot of the logic for running a query is the same between different panel types. For example, gauge chart and stat chart both call `useSuggestedStepMs` and `useTimeSeriesQuery`. This will lead to code duplications when we start supporting many different panel types.
3. We're giving panel plugin authors lots of flexibility when designing a panel plugin which isn't always good. 
4. This also makes it complicated to have consistent error and loading states across all panels because the data querying happens within the panel plugin (not at the panel chrome level). 
5. Right now there isn't a way to use visualization without embedding it in a panel.

**GOAL**
Ideally we should decouple data querying from panel plugin for the following reasons:

- Perses as a framework can then automatically determine which query editor component to show when user selects a panel type, offloading the data querying logic from panel plugin authors. 
- Developers can also reuse the data querying logic outside of a panel context. An example would be using a line chart without the panel header. 

**PROPOSAL**

1.  Remove `PanelQueryEditorComponent` and data querying logic from panel plugin. Instead, the panel plugin could specify which DataResponse formats it supports and that way Perses can automatically render the correct query editor component. This approach is similar to the approach Grafana takes. Grafana assumes every panel will also receive a DataFrame but gives you the ability to disable query editing with a declarative config option `skipDataQuery` if you don't use it. 

```
export interface PanelPlugin<Spec = UnknownSpec> extends Plugin<Spec> {
  PanelComponent: React.ComponentType<PanelProps<Spec>>;
  panelOptionsEditorComponents?: Array<PanelOptionsEditorComponent<Spec>>;
  supportedQueryInterfaces?: Array<'TimeSeriesQuery' | 'LogQuery' | 'TraceQuery'>; // if undefined, we can assume the panel does not need a query editor component
}
```

2. Add a hook that handles data querying and returns data in a format that the panel plugin understands and knows how to render it. Panel plugin can call this hook inside the panel component. 

```
type DataType = 'TimeSeries' | 'Logs' | 'Traces';

export function useData(dataType: DataType ): DataResponse<DataType> {
  //  add logic to query data  
}
```

3. Provide a way to use visualization without embedding it in a panel.

An idea from @saminzadeh:
```
<DataQueryProvider type="PrometheusTimeSeriesQuery" query={{query}} > 
  <TimeSeriesChart legend={{ position: "Right" }} /> 
</DataQueryProvider>
```

**Comment by sjcobb** [comment]
This make a lot of sense and I like how we can explicitly define which panel kinds work with which queries. Are any breaking changes in our dashboard definition for panel or query plugins being proposed?

**Reply by eunicorn** [reply]
If i understood correctly, in the weekly sync we talked about moving `queries` up a level to `PanelSpec`. 

**What we have now**
We currently define the queries as part of the chart options.

```
export interface TimeSeriesChartOptions {
  queries: TimeSeriesQueryDefinition[]; // we want to move this out of options
  legend?: LegendOptions;
  y_axis?: YAxisOptions;
  unit?: UnitOptions;
  thresholds?: ThresholdOptions;
  visual?: VisualOptions;
}
```

**Proposal**
We define the queries as part of the panel definition.
```
export interface PanelSpec<PluginSpec> {
  display: Display;
  queries: TimeSeriesQueryDefinition[]; // I am guessing we will need to update this to be more generic once we support other types of queries?
  plugin: Definition<PluginSpec>;
}
```

**Reply by LukeTillman** [reply]
Yeah, I think that's correct--the panel spec would have to change as well. Without a change like that, there's really no way for the framework code that's doing the querying for you to know what queries it should run.

**Comment by sjcobb** [comment]
_I misunderstood what was being proposed, as discussed on our weekly sync meeting, only the useTimeSeriesQueries lines would move out of TimeSeriesChartPanel_

~~Inside the current [query results loop](https://github.com/perses/perses/blob/v0.22.0/ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx#L137-L183), there is mapping for series display, legend data, and thresholds. Would this more generic `useData` hook mean an extra loop is needed in the panel for this customization?~~

~~If so, for performance to avoid the additional loop, we should probably convert our charts to use ECharts [dataset](https://echarts.apache.org/handbook/en/concepts/dataset/#define-data-in-dataset) so visual display is decoupled from raw data (downside is it might not give us the same level of customization as before). This would be a significant amount of work in our three chart panels and custom legend / tooltip, but overall is probably a better approach as we add more panel types.~~

**Comment by LukeTillman** [comment]
I mentioned this in our sync this morning, but I think querying is already basically "decoupled" from panel plugins and can be used independently. So long as you have the plugins available in a `PluginRegistry` component in your tree, you can just do the same thing that panel plugins do right now and call [`useTimeSeriesQuery`](https://github.com/perses/perses/blob/e66d205229622aa75d1823f08ab0bdbb1db8b737/ui/plugin-system/src/runtime/time-series-queries.ts#L79) or the variant that accepts multiple queries. You could then, for example, pipe that time series data to one of our vanilla visualization components from `@perses-dev/components` without needing `Panel`, etc.

It feels to me (and I could be totally wrong! 😂 ) like the other things you mention are maybe the real issues we're trying to fix here:

1. Ability to have standard loading/error states in a panel
2. Ability for the framework code to know about what queries are being run (and maybe even control when they actually execute, know when they error out, etc)
3. Make the editing experience more consistent by, for example, showing a query tab with the appropriate editor without plugin authors having to provide one themselves.

Does that seem right? What am I missing?

**Reply by saminzadeh** [reply]
> I mentioned this in our sync this morning, but I think querying is already basically "decoupled" from panel plugins and can be used independently. 

I agree that querying is decoupled from the panel plugin, but it's not entirely accurate to say that the panel plugins are decoupled from data querying. Data querying is only decoupled through a helper framework utility, so in this context, it can be considered coupled because the Panel plugin is responsible for integrating and implementing it within the plugin implementation. As a result, the Panel plugin cannot function without the data querying aspect.

> You could then, for example, pipe that time series data to one of our vanilla visualization components from @perses-dev/components without needing

That's true, however there is nothing to help enforce that interface separation. Components may be too low level and not "speak" the `TimeSeriesQuery` interface, making that more of a best practice. Our TimeSeriesPanel versus LineChart is a good illustration/example. We [implement a lot of logic ](https://github.com/perses/perses/blob/main/ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx#L89-L241)in the panel plugin and transform the data into props the [LineChart](https://github.com/perses/perses/blob/e66d205229622aa75d1823f08ab0bdbb1db8b737/ui/components/src/LineChart/LineChart.tsx) understands. The logic that transforms the TimeSeriesQuery result into props of a LineChart is coupled with the querying logic so they can't be used independently without pulling out some more utilities or refactoring. So while its possible, it's more difficult to enforce.

To summarize, I think it's better to handle some of these concerns at a framework level so we're not putting that responsibility on the plugin developer and enforcing that separation through a more restrictive plugin API interface.

> It feels to me (and I could be totally wrong! 😂 ) like the other things you mention are maybe the real issues we're trying to fix here:

Agreed 👍

**Reply by LukeTillman** [reply]
> That's true, however there is nothing to help enforce that interface separation. Components may be too low level and not "speak" the TimeSeriesQuery interface, making that more of a best practice. Our TimeSeriesPanel versus LineChart is a good illustration/example. We [implement a lot of logic ](https://github.com/perses/perses/blob/main/ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx#L89-L241)in the panel plugin and transform the data into props the [LineChart](https://github.com/perses/perses/blob/e66d205229622aa75d1823f08ab0bdbb1db8b737/ui/components/src/LineChart/LineChart.tsx) understands. The logic that transforms the TimeSeriesQuery result into props of a LineChart is coupled with the querying logic so they can't be used independently without pulling out some more utilities or refactoring. So while its possible, it's more difficult to enforce.

Do you think we should also be trying to make those vanilla components (like `LineChart`) "speak TimeSeriesQuery"? So maybe the props of that component take data in that shape and then the transformation code moves to the chart component instead of the panel?

**Reply by saminzadeh** [reply]
> Do you think we should also be trying to make those vanilla components (like LineChart) "speak TimeSeriesQuery"?

It would definitely help but I don't think it completely would solve the problem. While its a good idea, it's more of a best practice and hard to enforce. Those components are "dumb" and not Dashboard context aware (eg. time range, variable, annotations in the future). Ideally, you also don't have to implement features like onZoom <-> timeRange control and get those for free through standard context/providers. Happy to hop on a call to discuss more in detail of other use cases.

**Comment by eunicorn** [comment]
Let me just rephrase everything that's been mentioned:

So i think we have two major use cases that we need to consider here: 

**Use Case 1: Dashboard, Panel, Chart and how these all should work together** 

we want...
a. Ability to have standard loading/error states in a panel
b. Ability for the framework code to know about what queries are being run (and maybe even control when they actually execute, know when they error out, etc)
c. Make the editing experience more consistent by, for example, showing a query tab with the appropriate editor without plugin authors having to provide one themselves.
d. Ability to switch different panel types without re-entering the query each time

To solve issues A and B, we shouldn't run the queries inside the panel plugin. 

How/Proposed Solution: 
1. Define queries in PanelSpec instead of ChartOptions so the framework knows what queries to run. 
2. Implement a `DataQueryProvider` that accepts query type and queries and will run the queries.
3. Implement a hook `useData` that will get the data from DataQueryProvider 
4. Panel and PanelComponent can be children of DataQueryProvider so both Panel and PanelComponent have access to the data, error and loading states.

For example: in [GridItemContent.tsx](https://github.com/perses/perses/blob/main/ui/dashboards/src/components/GridLayout/GridItemContent.tsx)
```
<DataQueryProvider queryType={'TimeSeriesQuery} query={'sum(up)'}>
  <Panel definition={panelDefinition} />
</DataQueryProvider>
```
```
export function DataQueryProvider({queryType, queries, children}) {
    if (queryType === 'TimeSeriesQuery') {
        const {data, error, isLoading} = useTimeSeriesQuery(query, { suggestedStepMs });  
        return <DataQueryContext.Provider value={ctx}>{children}</DataQueryContext.Provider>
    }
}

export function useData() {
  return useContext(DataQueryContext);
}
```

In the future, we might even want to make `Panel` more composable by taking in children instead of panelDefinition.
```
<DataQueryProvider queryType={'TimeSeriesQuery} query={'sum(up)'}>
  <Panel>
      <PanelComponent />
  </Panel
</DataQueryProvider>
```

To solve issues C and D,  we can remove `PanelQueryEditorComponent` from the PanelPlugin definition and add `supportedQueryInterfaces` instead where each panel plugin will state what query editor it supports. This field is optional. If omitted, we can assume that a query editor is not needed. 

```
export interface PanelPlugin<Spec = UnknownSpec> extends Plugin<Spec> {
  PanelComponent: React.ComponentType<PanelProps<Spec>>;
  panelOptionsEditorComponents?: Array<PanelOptionsEditorComponent<Spec>>;
  supportedQueryInterfaces?: Array<'TimeSeriesQuery' | 'LogQuery' | 'TraceQuery'>; // if undefined, we can assume the panel does not need a query editor component
}
```

**Use Case 2: Using a chart without knowing anything about dashboard and panel**

Here we can wrap any charts with `DataQueryProvider` that is mentioned earlier.

```
<DataQueryProvider type="PrometheusTimeSeriesQuery" query={{query}} > 
  <TimeSeriesChart />
  <GaugeChart /> 
</DataQueryProvider>
```

**Reply by sjcobb** [reply]
A really good summary! To make this happen, will StatChart and GaugeChart panels need to be updated now to work with multiple queries or could that happen later?

**Reply by eunicorn** [reply]
i think that can come later, we can just add a condition in DataQueryProvider to call either useTimeSeriesQuery or useTimeSeriesQueries
